circuits/types/field/src/lsb.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use snarkvm_utilities::{FromBytes, ToBytes};
+use std::iter;
+
+impl<E: Environment> LSB for Field<E> {
+    type Boolean = Boolean<E>;
+
+    /// Returns the LSB of the integer.
+    fn lsb(&self) -> Self::Boolean {
+        let mut bytes = match self.eject_value().to_bytes_le() {
+            Ok(bytes) => bytes,
+            Err(_) => E::halt("Malformed field element detected when attempting to extract LSB."),
+        };
+
+        if self.is_constant() {
+            // Directly return the least significant bit.
+            Self::Boolean::new(Mode::Constant, bytes[0] & 1u8 == 1u8)
+        } else {
+            bytes.reverse();
+            // Take the byte containing the least significant bit.
+            let lsb_byte = match bytes.pop() {
+                Some(byte) => byte,
+                None => E::halt("Expected at least one byte."),
+            };
+            // Mask all but the lowest bit.
+            let mut lsb_bytes = vec![lsb_byte & 1u8];
+            // Zero-extend the field element.
+            lsb_bytes.extend(iter::repeat(0u8).take(bytes.len()));
+
+            // Mask lower bit.
+            bytes.push(lsb_byte & 254u8);
+            bytes.reverse();
+
+            // Instantiate witnesses for the lsb and upper bits of self.
+            let lsb = &Self::new(self.eject_mode(), E::BaseField::from_bytes_le(&lsb_bytes).unwrap());
+            let upper_bits = Self::new(self.eject_mode(), E::BaseField::from_bytes_le(&bytes).unwrap());
+
+            // Ensure (1 - lsb) * lsb = 0.
+            // `lsb` must be either 0 or 1.
+            E::assert_eq((Self::one() - lsb) * lsb, Self::zero());
+
+            // Ensure that the witnesses for the lsb and upper bits sum up to the original value.
+            E::assert_eq(self, upper_bits + lsb);
+
+            lsb.is_equal(&Self::one())
```

Comment:
This is unnecessary; since `lsb` is already boolean constrained, we should be able to convert the underlying low-level variable to the boolean type directly.

Reply:
Yes, this is castable for free